File: 9-1.py
#!/usr/bin/env python

input = [row.strip().split() for row in open('9.input').readlines()]

visited = set()

head = (0, 0)
tail = (0, 0)

for insn in input:
    dir = insn[0]
    steps = int(insn[1])

    for _ in range(steps):
        prev_head = head

        if dir == "U":
            head = (head[0], head[1] + 1)
        elif dir == "D":
            head = (head[0], head[1] - 1)
        elif dir == "L":
            head = (head[0] - 1, head[1])
        elif dir == "R":
            head = (head[0] + 1, head[1])

        if abs(head[0] - tail[0]) >= 2 or \
           abs(head[1] - tail[1]) >= 2:
            tail = prev_head

        visited.add(tail)

print(len(visited))

File: 9-1.rb
#!/usr/bin/env ruby

require 'set'

input = File.read('9.input').lines.map(&:strip).map(&:split)

visited = Set.new

head_x = 0
head_y = 0
tail_x = 0
tail_y = 0

input.each do |insn|
  dir = insn[0]
  steps = insn[1].to_i

  steps.times do
    prev_head_x = head_x
    prev_head_y = head_y

    case dir
    when 'U'
      head_y += 1
    when 'D'
      head_y -= 1
    when 'L'
      head_x -= 1
    when 'R'
      head_x += 1
    end

    if (head_x - tail_x).abs >= 2 or
       (head_y - tail_y).abs >= 2
      tail_x = prev_head_x
      tail_y = prev_head_y
    end

    visited.add [tail_x, tail_y]
  end
end

print visited.size, "\n"
